File: 2019/2019W24/preppindataw24.R
library(readxl)
library(dplyr)
library(stringr)
library(lubridate)

final <- read_excel("E:/Messages Input.xlsx", 
                   sheet = "chat") %>%
  mutate('Name' = str_extract(`Field_1`,'([:alpha:]+\\s*)+(?=\\W)'),
         'Message' = str_extract(`Field_1`,'(?<=:\\s).*'),
         'Number of Words' = sapply(strsplit(`Message`,' '),length),
         'Date' = parse_date_time(str_extract(`Field_1`,'\\d+/\\d+/\\d+'),'d/m/Y'),
         'Hour' = as.integer(str_extract(`Field_1`,'(?<=\\s)\\d+(?=:)'))) %>%
  merge(.,calendar <- read_excel("E:/Dates Input.xlsx", 
                                 sheet = "Dates") %>%
          mutate('Date'= parse_date_time(paste(`Date`,'2019'),'d b Y')), by='Date') %>%
  mutate('Send from work' = ifelse(`Holiday?`=='Weekday' &
                                      ((`Hour`>=9 & `Hour`<12) | (`Hour`>13 & `Hour`<17)),1,0)) %>%
  group_by(`Name`) %>%
  summarise('Text' = n(),
            'Avg Words/Sentence'=mean(`Number of Words`),
            'Number of Words'=sum(`Number of Words`),
            'Text While at Work'=sum(`Send from work`),
            '% Send from Work'=mean(`Send from work`)*100)

View(final)

File: 2019/2019W24/PreppindataW24.py
import pandas as pd
import re
from datetime import datetime as dt

calendar = pd.read_excel(pd.ExcelFile(r"E:\Dates Input.xlsx"),"Dates") 
calendar['Date']=calendar['Date'].apply(lambda x: dt.strptime(x+' 2019','%d %B %Y').date())

chat = pd.read_excel(pd.ExcelFile(r"E:\Messages Input.xlsx"),"chat")
chat['Name']=chat['Field_1'].apply(lambda x: re.search(']\s+\W*((\w+\s*)+)\W', x).group(1))
chat['Message']=chat['Field_1'].apply(lambda x: re.search(':\s+\W*(.*)', x).group(1))
chat['Number of Words']=chat['Message'].apply(lambda x:  len(x.split()))
chat['Date']=chat['Field_1'].apply(lambda x: dt.strptime(re.search('(\d+/\d+/\d+)', x).group(1),'%d/%m/%Y').date())
chat['Hour']=chat['Field_1'].apply(lambda x: int(re.search('\s(\d+):', x).group(1)))

chatdata = pd.merge(chat, calendar, how='inner', on='Date')
chatdata['Send from work']=chatdata.apply(lambda x: 1 if x['Holiday?']=='Weekday' and (9<=x['Hour']<12 or 13<x['Hour']<17) else 0,axis=1)
final = chatdata.groupby('Name',as_index=False).agg({'Number of Words':[('Number of Words', 'sum'), ('Avg Words/Sentence', 'mean')],
                     'Send from work':[('Text While at Work', 'sum'), ('% Send from Work', 'mean')],
                     'Message':[('Text','count')]})
final.columns = [col[1] if col[1] else col[0] for col in final.columns.values]
final['% Send from Work']=final['% Send from Work']*100
